packages/runtime/test-runtime-utils/src/mocksForReconnection.ts

```diff
@@ -91,68 +91,83 @@ export class MockContainerRuntimeForReconnection extends MockContainerRuntime {
+		const pendingMessages = fromContainerRuntime.pendingMessages.splice(0);
+		const remoteOps = [
+			...fromContainerRuntime.processedOps.splice(0),
+			...fromContainerRuntime.pendingRemoteMessages.splice(0),
+		];
 
 		// ensure no ops are sent to, or produced by the old client
 		// this can help find bugs in the the harness
-		Object.freeze(containerRuntime.pendingMessages);
-		Object.freeze(containerRuntime.pendingRemoteMessages);
-		Object.freeze(containerRuntime.processedOps);
-
-		// clear any unprocessed ops for this client
-		this.factory.clearOutstandingClientMessages(containerRuntime.clientId);
+		Object.freeze(fromContainerRuntime.pendingMessages);
+		Object.freeze(fromContainerRuntime.pendingRemoteMessages);
+		Object.freeze(fromContainerRuntime.processedOps);
 
 		let refSeq = Math.min(
-			savedOps[0]?.referenceSequenceNumber ?? Number.MAX_SAFE_INTEGER,
+			remoteOps[0]?.referenceSequenceNumber ?? Number.MAX_SAFE_INTEGER,
 			pendingMessages[0]?.referenceSequenceNumber ?? Number.MAX_SAFE_INTEGER,
 		);
 		if (refSeq === Number.MAX_SAFE_INTEGER) {
 			refSeq = 0;
 		}
-		this.dataStoreRuntime.deltaManager.lastSequenceNumber = refSeq;
-		this.dataStoreRuntime.deltaManager.minimumSequenceNumber = refSeq;
-
-		// handle pending messages before first saved op
-		while (
-			pendingMessages.length > 0 &&
-			pendingMessages[0].referenceSequenceNumber <=
-				this.dataStoreRuntime.deltaManager.lastSequenceNumber
+		if (
+			this.dataStoreRuntime.deltaManager.lastSequenceNumber !== refSeq ||
+			this.dataStoreRuntime.deltaManager.minimumSequenceNumber !== refSeq
 		) {
-			await this.dataStoreRuntime.applyStashedOp(pendingMessages.shift()?.content);
+			throw new Error(
+				"computed min and ref seq don't match the loaded values; this indicates a bad load, or missing messages",
+			);
 		}
-		// apply the saved and pending ops
-		for (const savedOp of savedOps) {
-			if (savedOp.clientId === this.clientId) {
-				await this.dataStoreRuntime.applyStashedOp(savedOp.contents);
+
+		const stashedOps = new Map<number, any>();
+
+		remoteOps.forEach((op) => {
+			if (op.clientId === this.clientId) {
+				const ops = stashedOps.get(op.referenceSequenceNumber) ?? [];
+				ops.push(op.contents);
+				stashedOps.set(op.referenceSequenceNumber, ops);
 			}
-			this.process(savedOp);
+		});
+		pendingMessages.forEach((op) => {
+			const ops = stashedOps.get(op.referenceSequenceNumber) ?? [];
+			ops.push(op.content);
+			stashedOps.set(op.referenceSequenceNumber, ops);
+		});
 
-			while (
-				pendingMessages.length > 0 &&
-				pendingMessages[0].referenceSequenceNumber === savedOp.sequenceNumber
-			) {
-				await this.dataStoreRuntime.applyStashedOp(pendingMessages.shift()?.content);
+		const applyStashedOpsAtSeq = async (seq: number) => {
+			const pendingAtSeq = stashedOps.get(seq);
+			while (pendingAtSeq && pendingAtSeq.length > 0) {
+				const message = pendingAtSeq.shift();
+				await this.dataStoreRuntime.applyStashedOp(message);
 			}
```

Comment:
nit: I generally prefer non-destructive versions in test code, since it makes reasoning about current state easier if needing to inspect locals up/down the stack frame
```suggestion
			for (const message of pendingAtSeq ?? []) {
				await this.dataStoreRuntime.applyStashedOp(message);
			}
```